pkg/codec/byte_string_modifier.go

```diff
@@ -114,97 +105,92 @@ func (t *bytesToStringModifier) RetypeToOffChain(onChainType reflect.Type, itemT
 			if err != nil {
 				return nil, err
 			}
-
 			offChainType = reflect.ArrayOf(onChainType.Len(), elm)
 		}
 
-		t.onToOffChainType[onChainType] = offChainType
-		t.offToOnChainType[offChainType] = onChainType
-
-		return offChainType, nil
 	case reflect.Struct:
 		return t.getStructType(onChainType)
 	default:
 		return nil, fmt.Errorf("%w: cannot retype the kind %v", types.ErrInvalidType, onChainType.Kind())
 	}
+
+	t.onToOffChainType[onChainType] = offChainType
+	t.offToOnChainType[offChainType] = onChainType
+	return offChainType, nil
 }
 
 // TransformToOnChain uses the AddressModifier for string-to-address conversion.
-func (t *bytesToStringModifier) TransformToOnChain(offChainValue any, itemType string) (any, error) {
+func (t *bytesToStringModifier) TransformToOnChain(offChainValue any, _ string) (any, error) {
 	return transformWithMaps(offChainValue, t.offToOnChainType, t.fields, noop, stringToAddressHookForOnChain(t.modifier))
 }
 
 // TransformToOffChain uses the AddressModifier for address-to-string conversion.
-func (t *bytesToStringModifier) TransformToOffChain(onChainValue any, itemType string) (any, error) {
+func (t *bytesToStringModifier) TransformToOffChain(onChainValue any, _ string) (any, error) {
 	return transformWithMaps(onChainValue, t.onToOffChainType, t.fields,
-		addressTransformationAction[bool](t.modifier.Length()),
+		addressTransformationAction(t.modifier.Length()),
 		addressToStringHookForOffChain(t.modifier),
 	)
 }
 
 // addressTransformationAction performs conversions over the fields we want to modify.
 // It handles byte arrays, ensuring they are convertible to the expected length.
 // It then replaces the field in the map with the transformed value.
-func addressTransformationAction[T any](length int) mapAction[T] {
-	return func(em map[string]any, fieldName string, value T) error {
-		// Check if the field exists in the map
+func addressTransformationAction(length int) func(extractMap map[string]any, key string, _ bool) error {
+	return func(em map[string]any, fieldName string, _ bool) error {
 		if val, ok := em[fieldName]; ok {
-			// Use reflection to extract the underlying value
 			rVal := reflect.ValueOf(val)
-			if rVal.Kind() == reflect.Ptr {
-				// Dereference the pointer if necessary
+
+			if !rVal.IsValid() {
+				return fmt.Errorf("invalid value for field %s", fieldName)
+			}
+
+			if rVal.Kind() == reflect.Ptr && !rVal.IsNil() {
```

Comment:
is nil a valid value here, if not the check above should check for both is nil and isValid and return an error

Reply:
may panic if we put it in the above line. Makes sense to first check if  `rVal.Kind() == reflect.Ptr` before calling `IsNil`